src/Distributions.cpp

```diff
@@ -415,7 +415,7 @@ NumericMatrix C_Vec_WeightedDiscretePdf(NumericVector x, NumericMatrix data,
   for (int i = 0; i < nc; i++) {
     for (int k = 0; k < n; k++) {
       for (int j = 0; j < nr; j++) {
-        if (data(j, i) == x[k]) {
+        if (data(j) == x[k]) {
```

Comment:
C++ not my strong point, but since `data` is a `NumericVector` as well, shouln't it be `data[j]`?

Reply:
Checking with https://teuder.github.io/rcpp4everyone_en/080_vector.html#accessing-vector-elements that should be correct. I wonder what actually happens when using `data(j, i)` in this case 🤔

Reply:
still it would be better to have it consistent, since both `x` and `data` are of the same type

Reply:
Verification question: if we never have the statement `data(j) == x[k]` TRUE, then the `mat(k, i)` (pdf at a new time point not included in the original `x`) will be zero and that's what we want, right?

Reply:
Yup

Reply:
Would it make sense to speed this up by removing the third loop? You pretty much search for the index where the time/data points match every time you scan the vector `data`

Reply:
How would you remove the third loop? Can you demonstrate with R code?

Reply:
Was writing the previous response late and didn't think it through, so what I initially thought is not going to work. But I asked ChatGTP and got something similar: an optimization that reduces the time complexity from `O(nr * nc * n)` to `O(nr + nc * n)` by precomputing the indices of elements in the 'data' array using a hashmap (!):
```
#include <unordered_map>

// Assuming mat, pdf, data, and x are appropriately defined

// Create a hashmap to store the indices of elements in the 'data' array
std::unordered_map<int, int> dataIndices;
for (int j = 0; j < nr; ++j) {
    dataIndices[data[j]] = j;
}

// Use the hashmap to populate the 'mat' matrix efficiently
for (int i = 0; i < nc; ++i) {
    for (int k = 0; k < n; ++k) {
        auto dataIndexIterator = dataIndices.find(x[k]);
        if (dataIndexIterator != dataIndices.end()) {
            int j = dataIndexIterator->second;
            mat(k, i) = pdf(j, i);
        }
    }
}
```

Reply:
Heh, looks like your ChatGPT-foo is better than mine, I didn't get too far 😬  
I thought that using an ordered set should be the way to go based on how indices are, well, ordered, so surely using an ordered data structure has some sort of benefit?
But I also didn't think this through entirely because I think I still haven't wrapped my head around this fully.